### ISSUE TITLE
 extraction optimizations performance issue

### ISSUE THREAD

**Issue by stefanradziuk** [rocq-prover/rocq#17448]
#### Description of the problem

I've been trying to extract some proofs (with results in Type, so they don't get erased) to Haskell. I've ran into an issue where the extraction would take unexpectedly long (not sure how long precisely, I ended up killing it after 30 minutes). However, I found that breaking up the proof by moving some of its subgoals into lemmas made the extraction complete in around 10 seconds (https://github.com/stefanradziuk/WasmCert-Coq/commit/cf861cd2ceee5073a2a6934ff1672b716d170a29). I didn't make the lemmas any more general, I just copied the available assumptions and the subgoal verbatim and made them into a separate proof. OCaml extraction completes in around 1 second for both cases.

I haven't managed to get a self-contained reproduction case, but I managed to narrow it down slightly 
[here](https://github.com/stefanradziuk/WasmCert-Coq/blob/057bd16a974b2fe38f2ca5b95a379951e18d4a32/theories/type_checker_reflects_typing_bugs_manually_reduced.v#L516-L738): `lemma_1` extracts in 10 s and `lemma_1_using_aux`, a 'broken up' version of `lemma_1` extracts in 0.2 s. Both extract in 0.02 s with `Unset Extraction Optimize`.

`lemma_1` extracted with optimizations on: [lemma_1_opt.hs.txt](https://github.com/coq/coq/files/11113402/lemma_1_opt.hs.txt) / with optimizations off: [lemma_1_unopt.hs.txt](https://github.com/coq/coq/files/11113408/lemma_1_unopt.hs.txt)

After looking at this with perf, there seems to be an exponential number of calls to `kill_dummy` and `kill_dummy_hd`, which suggests this may be because of the many `let in` expressions in the extracted code. Here's the perf data (as .txt to make GitHub happy) [perf.data.txt](https://github.com/coq/coq/files/11113445/perf.data.txt) and the [flamegraph](https://user-images.githubusercontent.com/7904551/228893341-9eff2c94-4768-404d-83a1-8ceedf61d00f.svg).

Zulip thread: https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs-.26-plugin-devs/topic/Haskell.20extraction.20performance.20issue

#### Coq Version

Coq 8.13.2

**Comment by ppedrot** [comment]
Indeed, kill_dummy is potentially exponential in the number of nested let-bindings. I'll try to come up with a fix.
